### PR TITLE
 Fixed #30752 -- Allowed using ExceptionReporter subclass in error re…

### PR DESCRIPTION
**Pull request by pulysak**
…ports.

https://code.djangoproject.com/ticket/30752

### CODE REVIEW

**Comment by carltongibson** [comment]
Thanks for this. 

We don’t want to add a setting here. Rather, add a parameter to the technical 500 handler, and show using `functools.partial` when providing the `handler500` in the URLConf. 

Documenting ExceptionReporter, perhaps just with the get_traceback_data() method show is a good idea, and needed at this point.

**Comment by pulysak** [comment]
> Thanks for this.
> 
> We don’t want to add a setting here. Rather, add a parameter to the technical 500 handler, and show using `functools.partial` when providing the `handler500` in the URLConf.
> 
> Documenting ExceptionReporter, perhaps just with the get_traceback_data() method show is a good idea, and needed at this point.

Hi, thank you for answer!

But as i see `handler500 ` and `technical_500_response` are different things. The first is view which we can set in settings and which runs when `DEBUG = False`.
The second one runs when  `DEBUG = True`. And it called from `response_for_exception` and `handle_uncaught_exception`. So i don't think that we can use `functools.partial` when providing the `handler500` in the URLConf.



Am i wrong?

Thank you!

**Comment by carltongibson** [comment]
Yes, you’re right. (Weekend! 🙂) 

I shall have a think Monday and get back to you. (I’d like to avoid a setting here but...)

**Comment by pulysak** [comment]
> Yes, you’re right. (Weekend! 🙂)
> 
> I shall have a think Monday and get back to you. (I’d like to avoid a setting here but...)

If we really don't want to get reporter class from settings, we can get it only from request class as i do it now.

**Comment by carltongibson** [comment]
Yes... I’m still pondering this, but was thinking exactly that as I was taking the bins out just now. 🙂

Musing: how would I do this in functional-land, given that we’re using functions at this stage? I’d pass in the callables I needed, rather than hard coding them. 

So we could do that, but is it worth it? Arguably not. So a setting it would have to be. 

I just need a cycle to think about what it looks like. Been busy with 3.0, but it’s top of my list to close/resolve as many of these Error Reporting issues as we can here. 

Thank you for your inputs! 🙂

File: docs/ref/settings.txt [hunk 1]
```diff
@@ -1201,6 +1201,16 @@ Default: ``'utf-8'``
...
 
+.. setting:: DEFAULT_EXCEPTION_REPORTER
+
+``DEFAULT_EXCEPTION_REPORTER``
+-------------------------------------
+
+Default: ``'``:class:`django.views.debug.ExceptionReporter`\ ``'``
+
+Default exception reporter class to be used if none has been assigned to
+the :class:`~django.http.HttpRequest` instance yet.
```
Comment (carltongibson) [inline docs/ref/settings.txt]: This should link to the error-reporting How-To.

File: docs/releases/3.0.txt [hunk 2]
```diff
@@ -279,6 +279,13 @@ Logging
   ``django.views.debug.ExceptionReporter`` subclass to customize the traceback
   text sent to site :setting:`ADMINS` when :setting:`DEBUG` is ``False``.
 
+Error reporter class
+~~~~~~~~~~~~~~~~~~~~~~
+* Added the :setting:`DEFAULT_EXCEPTION_REPORTER` setting defaults
+  to ``django.views.debug.ExceptionReporter``. This setting allows providing
+  ``django.views.debug.ExceptionReporter`` subclass to customize
+  the traceback text in error reports.
```
Comment (carltongibson) [inline docs/releases/3.0.txt]: Move to `3.1.txt`. Link to error reporting _How To_

**Comment by pulysak** [comment]
Done

**Comment by carltongibson** [comment]
Hi @pulysak. Thanks for the update. This is on my short(er) list. It's DjangoCon next week so... 😬 but if not then straight after. 

Thanks for your work here! 🏅

**Comment by pulysak** [comment]
@carltongibson HI, i'am sorry for disturbing you, but what do you think about the PR?

**Comment by carltongibson** [comment]
Hi @pulysak. Sorry for the delay here. This is very much on my list for Django 3.1 but we have had a long-line of other things come in, so I've only just got back to it. Looking again now (so resolution in the new year). Thanks  again for the effort!

File: docs/howto/error-reporting.txt [hunk 3]
```diff
@@ -305,6 +305,63 @@ following attributes and methods:
...
+As with the filter class, you may control which exception reporter class to use
+within any given view by setting the ``HttpRequest``’s
+``exception_reporter_class`` attribute::
+
+    def my_view(request):
+        if request.user.is_authenticated:
+            request.exception_reporter_class = CustomExceptionReporter()
+        ...
+
+.. currentmodule:: django.views.debug
```
Comment (felixxm) [inline docs/howto/error-reporting.txt]: This is not necessary, please chop this and following blank lines.

File: docs/howto/error-reporting.txt [hunk 4]
```diff
@@ -305,6 +305,63 @@ following attributes and methods:
...
+        This is the main extension point for customizing exception reports, for
+        example::
+
+            from django.views.debug import ExceptionReporter
+
+
+            class CustomExceptionReporter(ExceptionReporter):
+                def get_traceback_data(self):
+                    data = super().get_traceback_data()
+                    # ... remove/add something here...
```
Comment (felixxm) [inline docs/howto/error-reporting.txt]: `here...` -> `here ...`

File: docs/howto/error-reporting.txt [hunk 5]
```diff
@@ -305,6 +305,63 @@ following attributes and methods:
         traceback frame. Sensitive values are replaced with
         :attr:`cleansed_substitute`.
 
+If you need to customize error reports beyond filtering you may specify a
```
Comment (felixxm) [inline docs/howto/error-reporting.txt]: Maybe `versionadded` before this section :thinking:

File: tests/view_tests/tests/test_debug.py [hunk 6]
```diff
@@ -249,6 +249,15 @@ def test_technical_404_converter_raise_404(self):
...
             self.assertContains(response, 'Page not found', status_code=404)
 
+    def test_exception_reporter_from_request(self):
+        response = self.client.get('/custom_reporter_class_view/')
+        self.assertContains(response, 'custom traceback text', status_code=500)
+
+    @override_settings(DEFAULT_EXCEPTION_REPORTER='view_tests.views.CustomExceptionReporter')
+    def test_exception_reporter_from_settings(self):
+        response = self.client.get('/raises500/')
+        self.assertContains(response, 'custom traceback text', status_code=500)
```
Comment (felixxm) [inline tests/view_tests/tests/test_debug.py]: We can add tests for `CustomExceptionReporter.get_traceback_text()`, e.g.
```python
    def test_exception_reporter_from_settings(self):
        response = self.client.get('/raises500/', HTTP_X_REQUESTED_WITH='XMLHttpRequest')
        self.assertContains(response, 'custom traceback text', status_code=500)
```
or we can remove unused `CustomExceptionReporter.get_traceback_text()`.

Comment (carltongibson) [reply tests/view_tests/tests/test_debug.py]: I'll remove it. It's not under test in significant way.

**Comment by carltongibson** [comment]
Thanks for the review @felixxm. I [pulled those in](https://github.com/django/django/compare/9dea0bbd790a85d0bd629adf47dff3c7ed0d086e..e731e85c986e96597cb365f5f91769e2eb74bf35).

File: docs/howto/error-reporting.txt (hunk 5 above)
Comment (carltongibson) [reply docs/howto/error-reporting.txt]: This is a good idea. It adds a good "New in ..." header.

**Comment by pope1ni** [comment]
Sorry I'm late to the party. Could we have a little follow up commit to mention `exception_reporter_class` in https://docs.djangoproject.com/en/dev/ref/request-response/#attributes-set-by-application-code please? Thanks! 🙏

**Comment by carltongibson** [comment]
Hey Nick. Good idea. `exception_reporter_filter` isn't there either, and should be too, so we can add them both. I'll do that.

**Pull request merged**